src/main/java/org/intellij/sdk/codesync/codeSyncSetup/S3FileUploader.java

```diff
@@ -0,0 +1,131 @@
+package org.intellij.sdk.codesync.codeSyncSetup;
+
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import org.intellij.sdk.codesync.CodeSyncLogger;
+import org.intellij.sdk.codesync.clients.CodeSyncClient;
+import org.intellij.sdk.codesync.exceptions.FileLockedError;
+import org.intellij.sdk.codesync.exceptions.InvalidYmlFileError;
+import org.intellij.sdk.codesync.exceptions.RequestError;
+import org.intellij.sdk.codesync.files.S3UploadQueueFile;
+import org.intellij.sdk.codesync.repoManagers.OriginalsRepoManager;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class S3FileUploader {
+    S3UploadQueueFile s3UploadQueueFile;
+    Integer failedCount;
+    Map<String, Object> failedFilePathsAndURLs = new HashMap<>();
+
+    public S3FileUploader (S3UploadQueueFile s3UploadQueueFile) {
+        this.s3UploadQueueFile = s3UploadQueueFile;
+        this.failedCount = this.s3UploadQueueFile.getFailedCount();
+    }
+    public S3FileUploader (String repoPath, String branch, Map<String, Object> filePathAndURLs) {
+        this.failedCount = 0;
+        this.s3UploadQueueFile = new S3UploadQueueFile(repoPath, branch, filePathAndURLs, failedCount);
+    }
+
+    public void saveURLs() throws InvalidYmlFileError, FileNotFoundException {
+        try {
+            this.s3UploadQueueFile.writeYml();
+        } catch (FileLockedError error) {
+            // This would never happen, but adding it for completeness.
+            CodeSyncLogger.error("File locked error while creating a new file for S3 file upload queue.");
+        }
+    }
+
+    /*
+    Pre-Process and validate yml data. return `true` of we are good to proceed, `false` otherwise.
+    */
+    public boolean isValid() {
+        if (!this.s3UploadQueueFile.hasFiles()){
```

Comment:
I think we should assert that a file contains valid data structure
Ref: https://github.com/codesyncapp/codesync-vs/blob/master/src/init/s3_uploader.ts#L172

Reply:
@Basit-qc `hasFiles` is checking all the required fields are present.